File: my-folder/0022-generate-parentheses/solution.kt
class Solution {
    fun generateParenthesis(n: Int): List<String> {
        val res = mutableListOf<String>()

        fun valid(s: String): Boolean {
            var open = 0
            for (c in s) {
                if (c == '(') open++ else open--
                if (open < 0) return false
            }
            return open == 0
        }

        fun dfs(s: String) {
            if (s.length == n * 2) {
                if (valid(s)) {
                    res.add(s)
                }
                return
            }

            dfs(s + "(")
            dfs(s + ")")
        }

        dfs("")
        return res
    }
}

File: my-folder/0022-generate-parentheses/solution.go
func generateParenthesis(n int) []string {
    stack := []string{}
    res := []string{}

    var backtrack func(open int, close int)
    backtrack = func(open int, close int) {
        if open == close && close == n {
            res = append(res, strings.Join(stack, ""))
            return
        }

        if open < n {
            stack = append(stack, "(")
            backtrack(open+1, close)
            stack = stack[:len(stack)-1]
        }

        if close < open {
            stack = append(stack, ")")
            backtrack(open, close+1)
            stack = stack[:len(stack)-1]
        }
    }

    backtrack(0, 0)
    return res

}
